server/encryptionkm/key_manager.go

```diff
@@ -14,41 +14,452 @@
+		masterKeyMeta:         masterKeyMeta,
+		helper:                helper,
+	}
+	// Load encryption keys from storage.
+	_, err = m.loadKeys()
+	if err != nil {
+		return nil, err
+	}
+	// Start periodic check for keys change and rotation key if needed.
+	go m.startBackgroundLoop(ctx, m.keysRevision)
+	return m, nil
+}
+
+func (m *KeyManager) startBackgroundLoop(ctx context.Context, revision int64) {
+	// Create new context for the loop.
+	loopCtx, loopCancel := context.WithCancel(ctx)
+	defer loopCancel()
+	// Setup key dictionary watcher
+	watcher := clientv3.NewWatcher(m.etcdClient)
+	defer watcher.Close()
+	watcherCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	watchChan := watcher.Watch(watcherCtx, EncryptionKeysPath, clientv3.WithRev(revision))
+	// Check data key rotation every min(dataKeyRotationPeriod, keyRotationCheckPeriod).
+	checkPeriod := m.dataKeyRotationPeriod
+	if keyRotationCheckPeriod < checkPeriod {
+		checkPeriod = keyRotationCheckPeriod
+	}
+	ticker := time.NewTicker(checkPeriod)
+	defer ticker.Stop()
+	// Loop
+	for {
+		select {
+		// Reload encryption keys updated by PD leader (could be ourselves).
+		case resp := <-watchChan:
+			if resp.Canceled {
+				// If the watcher failed, we rely solely on rotateKeyIfNeeded() to reload encryption keys.
+				log.Warn("encryption key watcher canceled")
+				continue
+			}
+			for _, event := range resp.Events {
+				if event.Type != mvccpb.PUT {
+					log.Warn("encryption keys is deleted unexpectedly")
+					continue
+				}
+				{
+					m.muUpdate.Lock()
+					m.loadKeysFromKV(event.Kv)
+					m.muUpdate.Unlock()
+				}
+			}
+			m.helper.eventAfterReload()
+			// Check data key rotation in case we are the PD leader.
+		case <-m.helper.tick(ticker):
+			m.muUpdate.Lock()
+			err := m.rotateKeyIfNeeded(false /*forceUpdate*/)
+			if err != nil {
+				log.Warn("fail to rotate data encryption key", zap.Error(err))
+			}
+			m.muUpdate.Unlock()
+			m.helper.eventAfterTicker()
+			// Server shutdown.
+		case <-loopCtx.Done():
+			log.Info("encryption key manager is closed")
+			return
+		}
+	}
+}
+
+func (m *KeyManager) loadKeysFromKV(
+	kv *mvccpb.KeyValue,
+) (*encryptionpb.KeyDictionary, error) {
+	// Sanity check if keys revision is in order.
+	// etcd docs indicates watcher event can be out of order:
+	// https://etcd.io/docs/v3.4.0/learning/api_guarantees/#isolation-level-and-consistency-of-replicas
+	if kv.ModRevision <= m.keysRevision {
+		return m.getKeys(), nil
+	}
+	keys, err := loadKeysFromKV(kv)
+	if err != nil {
+		return nil, err
+	}
+	m.keysRevision = kv.ModRevision
+	m.keys.Store(keys)
+	log.Info("reloaded encryption keys", zap.Int64("revision", kv.ModRevision))
+	return keys, nil
+}
+
+func (m *KeyManager) loadKeys() (keys *encryptionpb.KeyDictionary, err error) {
+	resp, err := etcdutil.EtcdKVGet(m.etcdClient, EncryptionKeysPath)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil || len(resp.Kvs) == 0 {
+		return nil, nil
```

Comment:
If `m.getKeys()` is not `nil` at this time, it means that an accidental deletion has occurred and `err` needs to be returned.

Reply:
Not sure if I understand correctly, but for a PD instance without enabling encryption, we could get empty result.

Reply:
I am referring to the situation equivalent to `log.Warn("encryption keys is deleted unexpectedly")`

Reply:
got it. I will add a test.